### ISSUE TITLE
Add content length check when handling http requests

### ISSUE THREAD

**Issue by pps83** [nodejs/http-parser#331]
I was writing some code and got stuck with cryptic error and couldn't figure out where it came from:

```
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: Parse Error
    at Error (native)
    at Socket.socketOnData (_http_client.js:361:20)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:177:18)
    at Socket.Readable.push (_stream_readable.js:135:10)
    at TCP.onread (net.js:542:20)
```

in entire backtrace all the code is internal to nodejs and it's not obvious how to figure out what the problem was, I couldn't even go to any of these files to see that code.
At first I thought it was a bug in nodejs, since I was able to reproduce it in fairly simple scenario, after googling I saw [others also had that same 'mysterious' trace](https://github.com/SphtKr/homebridge-zway/issues/38).
When I tried to simplify my code I noticed that at some point the bug would disappear, and after digging for a couple of minutes I figured out what was the problem. I had a test http server started using `http.createServer` and there it generated random junk based on requested filename (for testing).

Here's simple code that reproduces it:

```
var http = require('http');
var data = Buffer.alloc(1024);
var server = http.createServer(function (req, res) {
    var rx = req.url.match(/\/(\d+)\.(\w+)/);
    res.writeHead(200, {'Content-Length': 1024, 'Content-Type': 'video/mp4'});
    var i = 0, w = function(){
        // ====> bug is here: (rx[1]), it has to be (+rx[1])
        res.write(data.slice(i, Math.min(1024, i+(rx[1]))-i));
        i += +rx[1];
        if (i < 1024)
            setTimeout(w, 10);
        else
        {
            res.end();
            next();
        }
    };
    w();
});
server.listen(8083);

http.get('http://localhost:8083/100.mp4', function(res){
    var len = 0;
    res.on('data', (chunk) => len += chunk.length);
});

```

Basically because of that bug my test server would end up wiring more that 1024 bytes in HTTP body, which makes receiver on the other end parse subsequent bytes as if it was new HTTP message.
It would be helpful in debug mode to track if `req` had `Content-Length` set and verify that code ends message when it's exactly that size body written, or message aborts without completing the body.
Is there such thing as debug build in nodejs by the way, something like NDEBUG in C/C++?
The other point: perhaps on receiver side it shouldn't have tried to parse past the end of the HTTP reply, since 'connection: close' (is that ok according to HTTP spec)?

**Comment by indutny** [comment]
This is intended behavior of this protocol. Bytes after content are parsed as a next request.
